crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -0,0 +1,764 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::cluster_controller::scheduler::ObservedClusterState;
+use rand::prelude::IteratorRandom;
+use rand::{thread_rng, RngCore};
+use restate_bifrost::{Bifrost, BifrostAdmin};
+use restate_core::metadata_store::{MetadataStoreClient, Precondition, WriteError};
+use restate_core::{metadata, task_center, Metadata, MetadataWriter};
+use restate_types::logs::builder::{ChainBuilder, LogsBuilder};
+use restate_types::logs::metadata::{
+    Chain, LogletConfig, LogletParams, Logs, ProviderKind, SegmentIndex,
+};
+use restate_types::logs::{LogId, Lsn};
+use restate_types::metadata_store::keys::BIFROST_CONFIG_KEY;
+use restate_types::nodes_config::Role;
+use restate_types::partition_table::PartitionTable;
+use restate_types::replicated_loglet::{NodeSet, ReplicatedLogletParams, ReplicationProperty};
+use restate_types::{Version, Versioned};
+use std::collections::BTreeMap;
+use std::num::NonZeroU8;
+use tokio::task::JoinSet;
+use tracing::debug;
+
+#[derive(Debug)]
+enum LogsState {
+    Provisioning {
+        log_id: LogId,
+        provider_kind: ProviderKind,
+    },
+    Running {
+        configuration: Option<LogletConfiguration>,
```

Comment:
When will this be `None`? Is it truly optional, if we believe a log is operational?

Reply:
It's not really optional. But it is a technicality to be able to easily move out of a `Running` state and transitioning into another state. I'll add a comment explaining it.